librad/src/meta/identity.rs

```diff
@@ -0,0 +1,1091 @@
+        };
+        if removed {
+            Ok(())
+        } else {
+            Err(Error::KeyNotPresent)
+        }
+    }
+
+    /// Add all keys from another `Delegation` as belonging to the given user
+    pub fn add_user_keys(
+        &mut self,
+        user_keys: &Self,
+        user_revision: &Revision,
+    ) -> Result<(), Error> {
+        if let (Delegations::Users(keys), Delegations::Keys(user_keys)) = (self, user_keys) {
+            for k in user_keys.iter() {
+                keys.insert(k.clone(), user_revision.clone());
+            }
+            Ok(())
+        } else {
+            Err(Error::MismatchedDelegation)
+        }
+    }
+
+    /// Remove the given keys
+    pub fn remove_keys(&mut self, delegations: &Self) {
+        for k in delegations.keys() {
+            self.remove_key(k).ok();
+        }
+    }
+
+    /// Compute the required quorum
+    pub fn quorum(&self) -> usize {
+        match self {
+            Delegations::Keys(keys) => (keys.len() / 2) + 1,
+            Delegations::Users(users) => {
+                let mut unique_users = BTreeSet::new();
+                for u in users.values() {
+                    unique_users.insert(u.as_bytes());
+                }
+                (unique_users.len() / 2) + 1
+            },
+        }
+    }
+
+    /// Check that the given signatures match the required quorum
+    pub fn check_quorum(&self, signatures: &BTreeMap<PublicKey, Signature>) -> bool {
+        match self {
+            Delegations::Keys(keys) => {
+                let mut count = 0;
+                for k in signatures.keys() {
+                    if keys.contains(k) {
+                        count += 1;
+                    }
+                }
+                count >= self.quorum()
+            },
+            Delegations::Users(users) => {
+                let mut unique_signers = BTreeSet::new();
+                for k in signatures.keys() {
+                    users.get(k).map(|u| unique_signers.insert(u.as_bytes()));
+                }
+                unique_signers.len() >= self.quorum()
+            },
+        }
+    }
+}
+
+/// Type witness for a fully verified [`Doc`].
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Verified;
+
+/// Type witness for a [`Doc`] signed by a quorum of its delegations.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Quorum;
+
+/// Type witness for a [`Doc`] with verified signatures.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Signed;
+
+/// Type witness for an untrusted [`Doc`].
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Untrusted;
+
+/// Identity document
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+pub struct Doc<Status>
+where
+    Status: Clone,
+{
+    /// Verification status
+    status_marker: PhantomData<Status>,
+    /// Previous document in the identity history
+    replaces: Option<Revision>,
+    /// Document payload (application defined)
+    payload: JsonValue,
```

Comment:
Since `Doc` is serialized in canonical json the `T` is expected to be serializable as json anyway.

The application is not expected to use a `JsonValue`: it is expected to use the generic `pub fn payload<T>(&self) -> Result<T, Error>` accessor and get its `T`.

Or, for identity creation, to pass the `T` when the identity is built.

The `JsonValue` is there so that at the Radicle Link level we can deserialize an `Identity` regardless of the kind (user or project), without needing union types or caring about different types `T` at all.

If you think of this as a protocol, the `Doc` is an envelope and the `T` is the payload carried inside the envelope.
A normal way to way to implement protocol encodings is to make it possible for the "envelope" level to ignore the payload entirely and just pass it along.